### PR TITLE
Update to rangev3 to 0.4.0

### PR DESCRIPTION
**Pull request by h-2**
Updates range-v3 to 0.4.0

diverse set of changes, I recommend going by individual commits.

Possibly introduces an ICE in concatenated_sequence with GCC8, I am investigating that.

### CODE REVIEW

**Comment by marehr** [comment]
The dox test is failing:

```
1: b'Parsing file /include/seqan3/std/type_traits:51: warning: Compound std::type_identity is not documented.\n'
1: b'Generating docs for file /include/seqan3/std/type_traits:56: warning: Member type_identity_t (typedef) of namespace std is not documented.\n'
1: b'/include/seqan3/std/type_traits:56: warning: Member type_identity_t (typedef) of namespace std is not documented.\n'
1: b'/include/seqan3/std/type_traits:52: warning: Member type (typedef) of class std::type_identity is not documented.\n'
```

File: include/seqan3/std/type_traits [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+#pragma once
+
+#include <type_traits>
+
+#if __cplusplus <= 201810L // TODO feature_test_macro if available
+
+namespace std
+{
+
+template <typename t>
```
Comment (marehr) [inline include/seqan3/std/type_traits]: documentation is missing

File: include/seqan3/range/view/take.hpp [hunk 2]
```diff
@@ -261,17 +264,17 @@ class view_take : public ranges::view_interface<view_take<urng_t, exactly, or_th
     //!\brief The reference_type.
     using reference         = reference_t<urng_t>;
     //!\brief The const_reference type is equal to the reference type if the underlying range is const-iterable.
-    using const_reference   = std::conditional_t<const_iterable_concept<urng_t>, reference, void>;
+    using const_reference   = detail::transformation_trait_or_t<seqan3::reference<urng_t const>, void>;
```
Comment (marehr) [inline include/seqan3/range/view/take.hpp]: oh cool, that you also have some use cases for it :)

File: include/seqan3/range/view/take.hpp [hunk 3]
```diff
@@ -403,6 +406,18 @@ class view_take : public ranges::view_interface<view_take<urng_t, exactly, or_th
         ranges::copy(begin(), end(), ranges::back_inserter(ret));
         return ret;
     }
+
+    //!\overload
+    template <sequence_container_concept container_t>
+    operator container_t() const
+    //!\cond
+        requires const_iterable_concept<urng_t> && std::CommonReference<reference_t<container_t>, reference>
```
Comment (marehr) [inline include/seqan3/range/view/take.hpp]: maybe switch the arguments, s.t. it is clearer that both versions require the same, except that the const version needs `const_iterable_concept`

File: include/seqan3/range/view/take.hpp (hunk 2 above)
Comment (h-2) [reply include/seqan3/range/view/take.hpp]: yeah, it's quite useful!

Comment (h-2) [reply include/seqan3/range/view/take.hpp]: yeah, it's quite useful!

Comment (smehringer) [reply include/seqan3/range/view/take.hpp]: echo ...  :laughing:

File: test/unit/range/view/view_persist_test.cpp [hunk 4]
```diff
@@ -83,8 +85,8 @@ TEST(view_persist, wrap_temporary)
     // combinability
     auto v3 = std::string{"foo"} | view::persist | ranges::view::unique;
     EXPECT_EQ("fo", std::string(v3));
-    std::string v3b = std::string{"foo"} | view::persist | view::reverse | ranges::view::unique;
```
[the remark this answers is not in the document]
Comment (h-2) [reply test/unit/range/view/view_persist_test.cpp]: no, as the commit message says, something is severely broken with view::reverse

File: include/seqan3/range/view/persist.hpp [hunk 5]
```diff
@@ -201,8 +202,7 @@ class view_persist : public ranges::view_interface<view_persist<urng_t>>
     template <sequence_container_concept container_t>
```
Comment (smehringer) [inline include/seqan3/range/view/persist.hpp]: `seqan3::`sequence_container_concept ?

File: include/seqan3/search/fm_index/fm_index.hpp [hunk 6]
```diff
@@ -232,7 +232,11 @@ class fm_index
         // * sdsl construction currently only works for int_vector, std::string and char *, not ranges in general
         // uint8_t largest_char = 0;
         sdsl::int_vector<8> tmp_text(text.size());
-        ranges::copy(text | view::reverse | view::to_rank | view::transform([] (uint8_t const r) { return r + 1; }),
+
+        //TODO view::reverse is broken and can't be chained right now
```
Comment (smehringer) [inline include/seqan3/search/fm_index/fm_index.hpp]: Is there an issue? If yes reference it, if no, make an issue and reference it :smile:

File: include/seqan3/range/view/take.hpp [hunk 7]
```diff
@@ -84,14 +86,15 @@ class view_take : public ranges::view_interface<view_take<urng_t, exactly, or_th
     //!\brief The sentinel type is identical to that of the underlying range.
     using sentinel_type = sentinel_t<urng_t>;
 
-    //!\brief The iterator type inherits from the underlying type, but overwrite several operators.
-    class iterator_type : public inherited_iterator_base<iterator_type, iterator_t<urng_t>>
+    //!\brief The iterator type inherits from the underlying type, but overwrites several operators.
+    template <typename rng_t>
```
Comment (smehringer) [inline include/seqan3/range/view/take.hpp]: why do you need the extra template?

File: include/seqan3/range/view/take.hpp [hunk 8]
```diff
@@ -403,6 +406,18 @@ class view_take : public ranges::view_interface<view_take<urng_t, exactly, or_th
...
+
+    //!\overload
+    template <sequence_container_concept container_t>
+    operator container_t() const
+    //!\cond
+        requires std::CommonReference<reference_t<container_t>, reference> && const_iterable_concept<urng_t>
+    //!\endcond
+    {
+        container_t ret;
+        ranges::copy(cbegin(), cend(), ranges::back_inserter(ret));
```
Comment (smehringer) [inline include/seqan3/range/view/take.hpp]: Don't we use the std::back_inserter now?

File: include/seqan3/search/fm_index/fm_index.hpp (hunk 6 above)
Comment (h-2) [reply include/seqan3/search/fm_index/fm_index.hpp]: I am already working on the issue...

File: include/seqan3/range/view/take.hpp (hunk 7 above)
Comment (h-2) [reply include/seqan3/range/view/take.hpp]: Hm, not obvious from the usage?

File: include/seqan3/range/view/take.hpp (hunk 8 above)
Comment (h-2) [reply include/seqan3/range/view/take.hpp]: Well, it was alle copy'n'paste, but your right, I can change it for the new ones. In general #498 has not been merged, yet.

**Comment by marehr** [comment]
Hmm, did you open a followup issue with the reverse problem?

**Pull request merged**